### PR TITLE
WIP: tuntap creation and reading 

### PR DESCRIPTION
**Pull request by si74**
NOTE: 
The runCmd, interface creation, etc will be merged into forked tuntap package at a later point. The remainder of this PR (ARP-related code, has been merged into main). 

TODOs: 
- ~~clean up cruft on macb w/ old routes and tun iface [X] - abandoned~~
- ~tun/tap basic reading [X]~
- ~confirm tun/tap writing works [X]~
- ~ensure addrs/routes added correctly[X] - having trouble here~ 
- CLEANUP: modularize interface creation/adding addrs [X]
- have tun/tap ipaddr route code work for openbsd, windows, linux [ ]
- ~more efficient data-writing (allocate buffer for an entire frame and properly read input data, then append ARP header) [X]~
- ~ARP stuffs - marshal/unmarshal, internal arp cache, send ARP request [X]~
- open tcp conn to controller and read/receive data [ ]
  - handle ingress (read from tap/tun -> send via tcp conn)
  - handle egress (read from tcp/conn -> write to tap/tun)

### CODE REVIEW

**Comment by si74** [comment]
Note: Right now we're shelling out an iproute2 command for route-setting. But discovered that we can use rtnetlink and the netlink sockets to perhaps achieve some of this work. 

Take a look at this:  https://man7.org/linux/man-pages/man7/netlink.7.html

And at this: https://github.com/vishvananda/netlink

File: cmd/tuntap/main.go [hunk 1]
```diff
@@ -290,6 +335,23 @@ func main() {
 		}
 	}()
 
+	go func() {
```
Comment (si74) [inline cmd/tuntap/main.go:338]: @nicboul namely I'm realizing my first thought process about putting this in another goroutine isn't a good idea. Perhaps we should talk this over.

File: cmd/tuntap/main.go [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+	Netmask string
+	Addr    string
+	Subnet  string
+}
+
+// arpTable is concurrent-safe ARP hashmap matching string IPv4 address to MAC addresses
+var ArpTable sync.Map
+
+type ArpEntry struct {
+	Mac       net.HardwareAddr
```
Comment (si74) [inline cmd/tuntap/main.go:32]: add field for IP here as well

File: cmd/tuntap/main.go [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+
+// IPConfig contains routing and interface configuration information
+type IPConfig struct {
+	Netmask string
+	Addr    string
+	Subnet  string
+}
+
+// arpTable is concurrent-safe ARP hashmap matching string IPv4 address to MAC addresses
+var ArpTable sync.Map
```
Comment (si74) [inline cmd/tuntap/main.go:29]: key is IP

File: cmd/tuntap/main.go [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+	}
+
+	log.Println("starting the tun/tap listener...")
+	log.Printf("Note traffic must be sourced from IP of iface like so: ping -S %v %v\n", addr, exampleDst)
+
+	buffLen := 100
+	waitBuffLen := 1000
+	ingressBuffer := make(chan []byte, buffLen)
+	egressBuffer := make(chan []byte, buffLen)
+	egressWaitBuffer := make(chan []byte, waitBuffLen)
```
Comment (si74) [inline cmd/tuntap/main.go:215]: linked list or hashmap instead for "waiting on Arp reply" queue. Type should be something that points to our existing buffer

File: cmd/tuntap/main.go [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+			var ipv4 net.IP
+			ipv4 = rawIP
+			fmt.Println(ipv4.String())
+
+			// if mac not found in hashtable
+			// add entry and push to arpChannel
+			entry, ok := ArpTable.Load(ipv4.String())
+			if !ok {
+				ArpTable.Store(ipv4.String(), ArpEntry{Status: StatusReady, Timestamp: time.Now()})
+				egressWaitBuffer <- buff
```
Comment (si74) [inline cmd/tuntap/main.go:303]: sends to our new hashmap/linked list thing. also TODO - send out ARP response

File: cmd/tuntap/main.go [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+				Protocol: 1,
+				Checksum: 0xdead,
+				Src:      net.ParseIP("198.18.0.5"), Dst: net.ParseIP("198.18.0.1"),
+			}
+			buff, err := hdr.Marshal()
+			if err != nil {
+				fmt.Printf("unable to parse header: %v\n", err)
+				continue
+			}
+			ingressBuffer <- buff
```
Comment (si74) [inline cmd/tuntap/main.go:273]: can check if ARP or not and send appropriately

**Comment by si74** [comment]
@nicboul as per our chat: 

Brainstorming doc here: https://docs.google.com/drawings/d/1roLmnkihKfqFOXS_ZxdSqFfvWaGotA8RPmPBtRcBsck/edit

1. I'll create an arp.go with general functions in it
2. will finish back functionality relating to sending arp, waiting for response, etc
3. reconstitute code from this WIP branch into main PR
4. then we work together on ARP table cleanup

File: cmd/tuntap/arp.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+func (t *ArpTable) Get(IP string) error {
+	return nil
+}
+
+// Remove
+func (t *ArpTable) Remove(IP string) error {
+	return nil
+}
+
+// ARPQueue is a thread-safe doubly-linked list buffer of of IP addresses waiting for ARP responses.
```
Comment (nicboul) [inline cmd/tuntap/arp.go:56]: "of of"

Comment (si74) [reply cmd/tuntap/arp.go:85]: oopsies

File: cmd/tuntap/arp.go [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+		if entry.IP != ip {
+			e = e.Next()
+			count++
+			continue
+		}
+
+		// If there is a match, send out bytes
+		// TODO - cannot seem to properly parse this at all- endianness issue?
+		len := binary.BigEndian.Uint16(entry.buff[16:18])
```
Comment (si74) [inline cmd/tuntap/arp.go:189]: @nicboul this bit is just not working gahh

File: cmd/tuntap/util.go [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+	// add ethernet header
+	copy(buff[0:6], dstMAC)
+	copy(buff[6:12], srcMAC)
+	copy(buff[12:14], etherTypeIPV4)
+	copy(buff[14:14+hdr.Len], packetBuff)
+
+	//fmt.Println(binary.BigEndian.Uint16(buff[16:18]))
+	// TODO - tests are breaking b/c of this -> I can't seem to get the length from the buffered header irrespective of byte order i'm trying
+	//fmt.Println(binary.LittleEndian.Uint16(packetBuff[0:2]))
+	fmt.Println(binary.LittleEndian.Uint32(packetBuff[2:4]))
```
Comment (si74) [inline cmd/tuntap/util.go:49]: @nicboul this is where I'm generating a test frame and I can't tell if this is due to host byte order differing from network byte order or what. Basically i'd expect this value to be 20 but alas when I print it out, it is not (i.e. ipv4.HeaderLen)

File: cmd/tuntap/main.go [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+
+// IPConfig contains routing and interface configuration information
+type IPConfig struct {
+	Netmask string
+	Addr    string
+	Subnet  string
+}
+
+// generic wrapper to run command
+func runCmd(name string, cmd *exec.Cmd) error {
```
Comment (si74) [inline cmd/tuntap/main.go:27]: //TODO: this will be in forked tuntap package

File: cmd/tuntap/main.go [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+	if err := cmd.Wait(); err != nil {
+		return fmt.Errorf("%v command failed to complete with error: %v", name, err)
+	}
+
+	log.Printf("command finished successfully: %v\n", name)
+	return nil
+}
+
+// configureInterface execs two commands
+func configureInterface(osName, iface string, ipConfig *IPConfig) error {
```
Comment (si74) [inline cmd/tuntap/main.go:51]: //TODO: this will be in forked tuntap package

File: cmd/tuntap/main.go [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+	if err != nil {
+		log.Printf("failed to add route: %v\n", err)
+		return fmt.Errorf("unable to add route")
+	}
+
+	return nil
+}
+
+// validateInterfaceMacOs ensures that the interface should be of the format utun<Num>
+func validateInterface(osName, iface string) bool {
```
Comment (si74) [inline cmd/tuntap/main.go:85]: //TODO: this will be in forked tuntap package

**Comment by si74** [comment]
this code has been integrated into main go client so closing this

**Pull request closed without merging**